tools/freeze_requirements.py

```diff
@@ -0,0 +1,75 @@
+import os
+from concurrent.futures import ThreadPoolExecutor
+from subprocess import run
+from typing import Tuple
+
+from tools.lib import gitroot
+from tools.lib import ts_print as print
+
+
+def worker(cmd: Tuple[str, ...]) -> Tuple[str, int]:
+    cmd_pretty = " ".join(cmd)
+    print(f"+ {cmd_pretty}")
+    proc = run(cmd, capture_output=True)
+    rc = proc.returncode
+    print(f"+ {cmd_pretty} returned {rc}")
+    if rc != 0:
+        print(f"stdout: {proc.stdout.decode()}\nstderr: {proc.stderr.decode()}")
+    return cmd_pretty, rc
+
+
+def main() -> int:
+    os.chdir(gitroot())
+
+    base_cmd = (
+        "pip-compile",
+        "--no-header",
+        "--no-annotate",
+        "--allow-unsafe",
+        "-q",
+    )
+
+    with ThreadPoolExecutor(max_workers=2) as executor:
+        futures = (
+            #            executor.submit(
+            #                worker,
+            #                (
+            #                    *base_cmd,
+            #                    "requirements-base.txt",
+            #                    "-o",
+            #                    "requirements-frozen.txt",
+            #                ),
+            #            ),
+            executor.submit(
+                worker,
+                (
+                    *base_cmd,
+                    "requirements-dev.txt",
+                    "-o",
+                    "requirements-dev-only-frozen.txt",
+                ),
+            ),
+            #            executor.submit(
+            #                worker,
+            #                (
+            #                    *base_cmd,
+            #                    "requirements-base.txt",
+            #                    "requirements-dev.txt",
+            #                    "-o",
+            #                    "requirements-dev-frozen.txt",
+            #                ),
+            #            ),
+        )
+
+    rc = 0
+    for future in futures:
+        try:
+            cmd_pretty, rc = future.result()
+        except Exception as e:
```

Comment:
we should catch a more specific error here -- I believe the only possible ones are `OSError` (command not found)

Reply:
Yeah, but would there be any benefit to that in this case? We don't like, do any branching based on what kind of exception, we only care that there's an exception and print it out.

Reply:
the general rule of thumb is "always catch only what you expect" -- that way an unexpected error can be noticed and handled properly and it doesn't silently hide bugs (`AttributeError` / `TypeError` / etc. could be raised by an errant refactor)

Reply:
it'll bubble out of the future and give us a crash + stacktrace there I think

Reply:
ha, never mind you're right